zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -723,6 +913,459 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
+
+            let length = meta.query_advice(blk_hdr_rlp_len_calc, Rotation::cur());
+            cb.require_zero("length default value is zero", length);
+
+            cb.gate(and::expr([
+                not::expr(or::expr([
+                    meta.query_fixed(q_number, Rotation::cur()),
+                    meta.query_fixed(q_var_field_256, Rotation::cur()),
+                ])),
+                meta.query_selector(q_blk_hdr_rlp),
+            ]))
+        });
+
+        meta.create_gate(
+            "Block header RLP: leading zeros column is boolean",
+            |meta| {
+                let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+
+                let blk_hdr_is_leading_zero =
+                    meta.query_advice(blk_hdr_is_leading_zero, Rotation::cur());
+                cb.require_boolean(
+                    "blk_hdr_is_leading_zero is boolean",
+                    blk_hdr_is_leading_zero,
+                );
+
+                cb.gate(meta.query_selector(q_blk_hdr_rlp))
+            },
+        );
+
+        let blk_hdr_rlp_is_zero = IsZeroChip::configure(
+            meta,
+            |meta| meta.query_selector(q_blk_hdr_rlp),
+            |meta| meta.query_advice(blk_hdr_rlp, Rotation::cur()),
+            blk_hdr_rlp_inv,
+        );
+
+        let blk_hdr_rlp_length_is_zero = IsZeroChip::configure(
+            meta,
+            |meta| meta.query_selector(q_blk_hdr_rlp),
+            |meta| meta.query_advice(blk_hdr_rlp_len_calc, Rotation::cur()),
+            blk_hdr_rlp_len_calc_inv,
+        );
+
+        for q_field in [q_number, q_var_field_256] {
+            meta.create_gate("Block header RLP: leading zeros checks", |meta| {
+                let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+
+                let blk_hdr_rlp_cur = meta.query_advice(blk_hdr_rlp, Rotation::cur());
+                let blk_hdr_is_leading_zero_cur =
+                    meta.query_advice(blk_hdr_is_leading_zero, Rotation::cur());
+                let blk_hdr_is_leading_zero_prev =
+                    meta.query_advice(blk_hdr_is_leading_zero, Rotation::prev());
+
+                let q_field_cur = meta.query_fixed(q_field, Rotation::cur());
+                let q_field_prev = meta.query_fixed(q_field, Rotation::cur());
+
+                cb.require_zero("Leading zero is actually zero", blk_hdr_rlp_cur);
+                cb.require_equal(
+                    "Leading zeros must be continuous or we are at the begining of the field",
+                    1.expr(),
+                    or::expr([blk_hdr_is_leading_zero_prev, q_field_prev]),
+                );
+
+                cb.gate(and::expr([blk_hdr_is_leading_zero_cur, q_field_cur]))
+            });
+        }
+
+        // Covers a corner case where LSB leading zeros can be skipped.
+        // This can occur when `blk_hdr_is_leading_zero` is set to 0 wrongly (the actual
+        // byte value is non-zero)
```

Comment:
The prover could set `blk_hdr_is_leading_zero` to zero without the actual RLP byte being zero. That can skip valid (non-zero) leading bytes. This check captures the inconsistency so that the circuit fails.